examples/hangman.rs

```diff
@@ -0,0 +1,274 @@
+                println!("This was correct!");
+                all_guesses += &new_guess;
+                if secret.chars().all(|c| all_guesses.contains(c)) {
+                    break 'won;
+                }
+            } else {
+                println!("No that letter is not in the word");
+                break;
+            }
+        }
+        step(&mut hangman);
+    }
+    // Redraw the lines else they could be outdated.
+    draw_lines(&mut lines, secret, &all_guesses);
+    wl.set_speed(Speed::instant());
+    // Draw a circle
+    wl.arc_right(100.0, 360.0);
+    // Draw the left eye
+    teleport_relative(&mut wl, Point { x: 55.0, y: 40.0 });
+    wl.arc_right(5.0, 360.0);
+    // Draw the right eye
+    teleport_relative(&mut wl, Point { x: 85.0, y: 0.0 });
+    wl.arc_right(5.0, 360.0);
+    // Go to the start of the mouth
+    teleport_relative(&mut wl, Point { x: -85.0, y: -80.0 });
+    wl.right(90.0);
+    // Either draw it smiling or sad
+    if secret.chars().all(|c| all_guesses.contains(c)) {
+        println!("You Won!");
+        wl.right(45.0);
+        wl.arc_left(65.0, 90.0);
+    } else {
+        println!("You Lost!");
+        wl.left(45.0);
+        wl.arc_right(65.0, 90.0);
+    }
+}
+
+/// A function to ask the user for a letter
+fn ask_for_letter(guess: &mut String) -> Result<(), Error> {
+    print!("Please guess a letter: ");
+    std::io::stdout().flush()?;
+    std::io::stdin().read_line(guess)?;
+
+    // remove trailing newline characters
+    let to_trim = guess.trim_end_matches(&['\r', '\n'][..]).len();
+    guess.truncate(to_trim);
+
+    println!("You guessed: {}", &guess);
+    Ok(())
+}
+
+/// A function to conveniently teleport the turtle relative to its current position.
+fn teleport_relative(turtle: &mut Turtle, direction: Point) {
+    let cur = turtle.position();
+    let new: Point = cur + direction;
+    turtle.pen_up();
+    turtle.go_to(new);
+    turtle.pen_down();
+}
+
+/// Draw the lines one for each character. If the character has been guessed already correctly draw a green circle above the line.
+fn draw_lines(lines: &mut Turtle, secret: &str, all_guesses: &str) {
+    // hide the lines-turtle and go th the startigpoint of the lines.
+    lines.reset();
+    lines.hide();
+    lines.set_speed(Speed::instant());
+    lines.right(90.0);
+    teleport_relative(lines, Point { x: -300.0, y: 200.0 });
+
+    // draw a line for every character in the secret.
+    for letter in secret.chars() {
+        if all_guesses.contains(letter) {
+            teleport_relative(lines, Point { x: 2.5, y: 10.0 });
+
+            // as turtle is not yet able to draw text draw a green replacement circle.
+            lines.left(90.0);
+            lines.set_fill_color("green");
+            lines.set_pen_color("green");
+            lines.begin_fill();
+            lines.arc_right(7.5, 360.0);
+            lines.end_fill();
+            lines.set_pen_color("black");
+            lines.right(90.0);
```

Comment:
I don't suppose turtle returns itself so you could chain all these together like a builder pattern?

Reply:
no it does not (yet) but this could be a nice extension maybe?

Reply:
It does not, and given that this isn't building anything, I don't think it should. Every method call has a side-effect (similar to printing), so regular methods that don't return anything are appropriate. The builder pattern would make sense if we wanted to accumulate a bunch of information and then do something with it all at once at the end. That is not the case here.

Thanks for the suggestion though! :)